File: CSharp/CSharp/repository/UserDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CSharp.domain;
using CSharp.domain.validators;
using CSharp.Repository;
using log4net;

namespace CSharp.repository
{
    public class UserDBRepository : IUserRepository
    {
        private static readonly ILog log = LogManager.GetLogger("UserDBRepository");

        private readonly IValidator<User> userValidator;
        public UserDBRepository(IValidator<User> userValidator)
        {
            this.userValidator = userValidator;
            log.Info("Creating UserDBRepository");
        }

        public IEnumerable<User> FindAll()
        {
            IDbConnection con = JdbcManager.GetConnection();
            IList<User> users = new List<User>();

            using (var command = con.CreateCommand())
            {
                command.CommandText = "select id, username, password from Users";

                using (var dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        long idUser = dataReader.GetInt64(0);
                        string username = dataReader.GetString(1);
                        string password = dataReader.GetString(2);
                        User user = new User(username, password);
                        user.Id = idUser;
                        users.Add(user);
                    }
                }
            }
            log.Info(users);
            return users;
        }

        public void Save(User entity)
        {
            log.InfoFormat("Entered save with entity {0}", entity);
            IDbConnection con = JdbcManager.GetConnection();

            using (var command = con.CreateCommand())
            {
                command.CommandText = "insert into Users(username, password) values (@username, @password)";
                var paramUsername = command.CreateParameter();
                paramUsername.ParameterName = "@username";
                paramUsername.Value = entity.UserName;
                command.Parameters.Add(paramUsername);

                var paramPassword = command.CreateParameter();
                paramUsername.ParameterName = "@password";
                paramUsername.Value = entity.Password;
                command.Parameters.Add(paramPassword);

                var result = command.ExecuteNonQuery();
                if (result == 0)
                    Console.WriteLine("No user added!");
            }
        }


        public User FindOne(long id)
        {
            log.InfoFormat("Entering findOne with value {0}", id);

            IDbConnection con = JdbcManager.GetConnection();

            using (var command = con.CreateCommand())
            {
                command.CommandText = "select id, username, password from Users where id = @id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);

                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        long idUser = dataReader.GetInt64(0);
                        string username = dataReader.GetString(1);
                        string password = dataReader.GetString(2);
                        User user = new User(username, password);
                        user.Id = idUser;
                        log.InfoFormat("Exiting findOne with value {0}", user);
                        return user;
                    }
                }
            }
            log.InfoFormat("Exiting findOne with value {0}", null);

            return null;
        }

        public void Delete(long id)
        {
            log.InfoFormat("entering delete with id {0}", id);
            IDbConnection con = JdbcManager.GetConnection();
            using (var command = con.CreateCommand())
            {
                command.CommandText = "delete from Users where id = @id";
                var paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);

                var dataR = command.ExecuteNonQuery();
                if (dataR == 0)
                {
                    Console.WriteLine("No user deleted!");
                }
            }
        }

        public void Update(long id, User entity)
        {
            IDbConnection con = JdbcManager.GetConnection();
            using (var command = con.CreateCommand())
            {
                command.CommandText = "update Users set username = @username, password = @password where id = @id";
                var paramUsername = command.CreateParameter();
                paramUsername.ParameterName = "@username";
                paramUsername.Value = entity.UserName;
                command.Parameters.Add(paramUsername);

                var paramPassword = command.CreateParameter();
                paramUsername.ParameterName = "@password";
                paramUsername.Value = entity.Password;
                command.Parameters.Add(paramPassword);

                var paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;

                var dataR = command.ExecuteNonQuery();
                if (dataR == 0)
                {
                    Console.WriteLine("No user updated!");
                }
            }
        }

        public void Login(string username, string password)
        {
            log.InfoFormat("Entering Login with username {0}, password {1}", username, password);

            IDbConnection con = JdbcManager.GetConnection();

            using (var command = con.CreateCommand())
            {
                command.CommandText =
                    "select id, username, password from Users where username = @username and password = @password";
                IDbDataParameter paramUsername = command.CreateParameter();
                paramUsername.ParameterName = "@username";
                paramUsername.Value = username;
                command.Parameters.Add(paramUsername);

                command.CommandText =
                    "select id, username, password from Users where username = @username and password = @password";
                IDbDataParameter paramPassword = command.CreateParameter();
                paramPassword.ParameterName = "@password";
                paramPassword.Value = password;
                command.Parameters.Add(paramPassword);

                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        log.Info("Login successful");
                    }
                    else
                    {
                        log.Info("Invalid account information");
                        throw new RepositoryException("Invalid credentials");
                    }
                }
            }



        }
    }


}

File: MPP-Java/src/main/java/repository/UserDBRepository.java
package repository;

import domain.User;
import domain.validators.ValidationException;
import domain.validators.Validator;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class UserDBRepository implements UserRepository {

    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();
    private static Validator<User> validator;

    public UserDBRepository(Properties props, Validator<User> validator) {
        logger.info("Initializing UserDBRepository with properties: {} ", props);
        dbUtils = new JdbcUtils(props);
        UserDBRepository.validator = validator;
    }

    @Override
    public User login(String username, String password) throws RepositoryException {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = con.prepareStatement("select * from Users where username = ? and password = ?")) {
            preparedStatement.setString(1, username);
            preparedStatement.setString(2, password);
            try (ResultSet result = preparedStatement.executeQuery()) {
                if (result.next()) {
                    logger.traceExit();
                    return new User(username, password);
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.out.println(ex);
        }
        logger.traceExit("Username and password combination not valid");
        throw new RepositoryException("Invalid username or password!");
    }

    @Override
    public User findOne(Long aLong) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = con.prepareStatement("select * from Users where id=?")) {
            preparedStatement.setLong(1, aLong);
            try (ResultSet result = preparedStatement.executeQuery()) {
                if (result.next()) {
                    String username = result.getString("username");
                    String password = result.getString("password");
                    User user = new User(username, password);
                    logger.traceExit(user);
                    return user;
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.out.println(ex);
        }
        logger.traceExit("No user with id {} found", aLong);
        return null;
    }

    @Override
    public List<User> findAll() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<User> users = new ArrayList<>();
        try (PreparedStatement preparedStatement = con.prepareStatement("select * from Users")) {
            try (ResultSet result = preparedStatement.executeQuery()) {
                while (result.next()) {
                    Long id = (long) result.getInt("id");
                    String username = result.getString("username");
                    String password = result.getString("password");
                    User user = new User(username, password);
                    user.setId(id);
                    users.add(user);
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit(users);
        return users;

    }

    @Override
    public void save(User entity) throws ValidationException {
        logger.traceEntry("Saving user {}", entity);
        try{
            validator.validate(entity);
        }
        catch (ValidationException e) {
            throw e;
        }

        Connection con = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = con.prepareStatement("insert into Users(username,password) values (?,?)")) {
            preparedStatement.setString(1, entity.getUsername());
            preparedStatement.setString(2, entity.getPassword());
            int result = preparedStatement.executeUpdate();
            logger.trace("Saved {} instances", result);

        } catch (SQLException ex) {
            logger.error(ex);
            System.out.println("Invalid username\n");
        }

    }

    @Override
    public void delete(Long aLong) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("delete from Users where id=?")) {
            preStmt.setLong(1, aLong);
            int result = preStmt.executeUpdate();
            logger.trace("Deleted {} instances", result);
        } catch (SQLException ex) {
            logger.error(ex);
            System.out.println("Error DB " + ex);
        }
        logger.traceExit();

    }

    @Override
    public void update(Long aLong, User entity) throws ValidationException {
        logger.traceEntry("update user {}", entity);

        try{
            validator.validate(entity);
        }
        catch (ValidationException e) {
            throw e;
        }

        Connection con = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = con.prepareStatement("update Users set username = ?, password = ? where id = ?;")) {
            preparedStatement.setString(1, entity.getUsername());
            preparedStatement.setString(2, entity.getPassword());
            preparedStatement.setLong(3, aLong);
            int result = preparedStatement.executeUpdate();
            logger.trace("Updated {} instances", result);
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
    }
}
